File: frontend/loanflow-web/src/app/features/policy/services/policy.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  PolicyResponse,
  PolicyRequest,
  ApiResponse,
  PageResponse,
  PolicyStatsResponse
} from '../models/policy.model';

@Injectable({
  providedIn: 'root'
})
export class PolicyService {
  private apiUrl = '/api/v1/policies';

  constructor(private http: HttpClient) {}

  // ==================== CRUD ====================

  create(request: PolicyRequest): Observable<ApiResponse<PolicyResponse>> {
    return this.http.post<ApiResponse<PolicyResponse>>(this.apiUrl, request);
  }

  getById(id: string): Observable<ApiResponse<PolicyResponse>> {
    return this.http.get<ApiResponse<PolicyResponse>>(`${this.apiUrl}/${id}`);
  }

  getByCode(policyCode: string): Observable<ApiResponse<PolicyResponse>> {
    return this.http.get<ApiResponse<PolicyResponse>>(`${this.apiUrl}/code/${policyCode}`);
  }

  update(id: string, request: PolicyRequest): Observable<ApiResponse<PolicyResponse>> {
    return this.http.put<ApiResponse<PolicyResponse>>(`${this.apiUrl}/${id}`, request);
  }

  delete(id: string): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`);
  }

  // ==================== Listing & Search ====================

  list(page: number = 0, size: number = 20, sort: string = 'createdAt,desc'): Observable<ApiResponse<PageResponse<PolicyResponse>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);
    return this.http.get<ApiResponse<PageResponse<PolicyResponse>>>(this.apiUrl, { params });
  }

  listByCategory(category: string, page: number = 0, size: number = 20): Observable<ApiResponse<PageResponse<PolicyResponse>>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<ApiResponse<PageResponse<PolicyResponse>>>(`${this.apiUrl}/category/${category}`, { params });
  }

  search(query: string, page: number = 0, size: number = 20): Observable<ApiResponse<PageResponse<PolicyResponse>>> {
    const params = new HttpParams()
      .set('q', query)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<ApiResponse<PageResponse<PolicyResponse>>>(`${this.apiUrl}/search`, { params });
  }

  // ==================== Lifecycle ====================

  activate(id: string): Observable<ApiResponse<PolicyResponse>> {
    return this.http.patch<ApiResponse<PolicyResponse>>(`${this.apiUrl}/${id}/activate`, {});
  }

  deactivate(id: string): Observable<ApiResponse<PolicyResponse>> {
    return this.http.patch<ApiResponse<PolicyResponse>>(`${this.apiUrl}/${id}/deactivate`, {});
  }

  createNewVersion(id: string): Observable<ApiResponse<PolicyResponse>> {
    return this.http.post<ApiResponse<PolicyResponse>>(`${this.apiUrl}/${id}/versions`, {});
  }

  // ==================== Queries ====================

  getVersionHistory(policyCode: string): Observable<ApiResponse<PolicyResponse[]>> {
    return this.http.get<ApiResponse<PolicyResponse[]>>(`${this.apiUrl}/code/${policyCode}/versions`);
  }

  getActivePolicies(loanType: string): Observable<ApiResponse<PolicyResponse[]>> {
    return this.http.get<ApiResponse<PolicyResponse[]>>(`${this.apiUrl}/active/${loanType}`);
  }

  getStats(): Observable<ApiResponse<PolicyStatsResponse>> {
    return this.http.get<ApiResponse<PolicyStatsResponse>>(`${this.apiUrl}/stats`);
  }
}

File: backend/policy-service/src/main/java/com/loanflow/policy/service/PolicyService.java
package com.loanflow.policy.service;

import com.loanflow.dto.request.PolicyRequest;
import com.loanflow.dto.response.PolicyResponse;
import com.loanflow.policy.domain.enums.LoanType;
import com.loanflow.policy.domain.enums.PolicyCategory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface for policy operations
 */
public interface PolicyService {

    /**
     * Create a new policy
     */
    PolicyResponse create(PolicyRequest request, String createdBy);

    /**
     * Get policy by ID
     */
    PolicyResponse getById(String id);

    /**
     * Get the latest version of a policy by its code
     */
    PolicyResponse getByPolicyCode(String policyCode);

    /**
     * Get all versions of a policy
     */
    List<PolicyResponse> getVersionHistory(String policyCode);

    /**
     * Update a policy (only DRAFT or INACTIVE policies)
     */
    PolicyResponse update(String id, PolicyRequest request, String modifiedBy);

    /**
     * Create a new version of an existing policy
     */
    PolicyResponse createNewVersion(String id, String createdBy);

    /**
     * Activate a policy
     */
    PolicyResponse activate(String id, String modifiedBy);

    /**
     * Deactivate a policy
     */
    PolicyResponse deactivate(String id, String modifiedBy);

    /**
     * Delete a policy (only DRAFT policies)
     */
    void delete(String id);

    /**
     * List all policies with pagination
     */
    Page<PolicyResponse> listAll(Pageable pageable);

    /**
     * List policies by category
     */
    Page<PolicyResponse> listByCategory(PolicyCategory category, Pageable pageable);

    /**
     * Search policies by text
     */
    Page<PolicyResponse> search(String searchText, Pageable pageable);

    /**
     * Get active policies for a loan type (used by evaluation engine)
     */
    List<PolicyResponse> getActivePoliciesForLoanType(LoanType loanType);

    /**
     * Get active policies by category and loan type
     */
    List<PolicyResponse> getActivePoliciesByCategoryAndLoanType(PolicyCategory category, LoanType loanType);

    /**
     * Get policy statistics (count by category and status)
     */
    PolicyStatsResponse getStats();

    /**
     * Inner class for stats response
     */
    record PolicyStatsResponse(
            long totalPolicies,
            long activePolicies,
            long draftPolicies,
            long inactivePolicies,
            java.util.Map<String, Long> byCategory
    ) {}
}
